src/core/lpParser.mly

```diff
@@ -0,0 +1,417 @@
+%type <Syntax.p_module_path> path
+%type <Syntax.p_config> config
+%type <Syntax.qident> qident
+%type <Syntax.p_rw_patt> rw_patt_spec
+%type <Syntax.p_tactic list * Syntax.p_proof_end> proof
+
+// Precedences listed from low to high
+%nonassoc COMMA IN
+%right ARROW
+
+%%
+
+// An identifier with a boolean to true if it uses escaped syntax.
+ident: i=ID { (make_pos $loc (fst i), snd i) }
+// Identifier without the boolean
+any_ident: i=ID { make_pos $loc (fst i) }
+
+// A list of [ident] separated by dots
+path: ms=separated_nonempty_list(DOT, ID) { ms }
+
+// [path] with a post processing to yield a [qident]
+qident: p=path
+    {
+      let (mp, id) = List.split_last p in
+      make_pos $loc(p) (mp, fst id)
+    }
+
+term_ident:
+  | AT qid=qident { make_pos $loc (P_Iden(qid, true)) }
+  | qid=qident { make_pos $loc (P_Iden(qid, false)) }
+
+// Rewrite pattern identifier
+patt:
+  | DOLLAR p=ident { Some(fst p) }
+  | DOLLAR WILD { None }
+
+// Identifiers for arguments
+arg_ident:
+  | i=ident { Some(fst i) }
+  | WILD { None }
+
+// Arguments of abstractions, products &c.
+arg:
+  // Explicit argument without type annotation
+  | x=arg_ident { ([x], None, false) }
+  // Explicit argument with type annotation
+  | L_PAREN xs=arg_ident+ COLON a=term R_PAREN
+      { (xs, Some(a), false) }
+  // Implicit argument (possibly with type annotation)
+  | L_CU_BRACKET xs=arg_ident+ a=preceded(COLON, term)? R_CU_BRACKET
+      { (xs, a, true) }
+
+// Patterns of the rewrite tactic
+rw_patt_spec:
+  | t=term { P_rw_Term(t) }
+  | IN t=term { P_rw_InTerm(t) }
+  | IN x=ident IN t=term { P_rw_InIdInTerm(fst x, t) }
+  | u=term IN x=term t=preceded(IN, term)?
+    {
+      let ident_of_term {elt; _} =
+        match elt with
+          | P_Iden({elt=([], x); pos}, _) -> Pos.make pos x
+          | _ -> $syntaxerror
+      in
+      match t with
+      | Some(t) -> P_rw_TermInIdInTerm(u, ident_of_term x, t)
+      | None -> P_rw_IdInTerm(ident_of_term u, x)
+    }
+  | u=term AS x=ident IN t=term { P_rw_TermAsIdInTerm(u, fst x, t) }
+
+// Rewrite tactic pattern with enclosing brackets.
+rw_patt: L_SQ_BRACKET r=rw_patt_spec R_SQ_BRACKET { make_pos $loc r }
+
+// Tactics available in proof mode.
+tactic:
+  | INTRO xs=arg_ident+ { make_pos $loc (P_tac_intro(xs)) }
+  | APPLY t=term { make_pos $loc (P_tac_apply(t)) }
+  | SIMPL { make_pos $loc P_tac_simpl }
+  | REWRITE l=ASSOC? p=rw_patt? t=term
+    {
+      let b =
+        match l with
+        | Some(Pratter.Left) -> false
+        | _ -> true
+      in
+      make_pos $loc (P_tac_rewrite(b,p,t))
+    }
+  | REFINE t=term { make_pos $loc (P_tac_refine(t)) }
+  | REFL { make_pos $loc P_tac_refl }
+  | SYMMETRY { make_pos $loc P_tac_sym }
+  | WHY3 s=STRINGLIT? { make_pos $loc (P_tac_why3(s)) }
+  | q=query { make_pos $loc (P_tac_query(q)) }
+  | FAIL { make_pos $loc P_tac_fail }
+
+// Terminated tactic. The semicolon isn't necessary, it is added for
+// uniformity.
+ttactic: t=tactic SEMICOLON? { t }
```

Comment:
It will be better to make semi-colons mandatory here, for https://github.com/Deducteam/lambdapi/issues/515, but this may be done in a separate PR.

Reply:
I think we can easily parse based on the tactics keywords. But it may be more homogeneous to use semicolons to end tactics as well.

Reply:
In the future, I hope, tacticals should be added, so between the tactics, it will be possible to have different separators.
Don't forget that.